(#4484) Add Audio-Described Link to Video Content Type

dev-rana-publicis opened the pull request:
Closes #4484
ODE: https://ncigovcdode871.prod.acquia-sites.com/
**NOTE:** This PR will be closed after #4522 is merged. Will be keeping this PR open to refer back to old reviews and discussions. Ticket #4484 will be merged in with PR #4522

andyvanavery31 commented:
Things to update:

- The display options without caption need to have the caption of the video removed, but it still needs the audio-described version to test when there is an AD link but no caption.

alinai commented:
Tool-tip text on ODE does not match the AC:
![image](https://github.com/user-attachments/assets/a015d250-0104-49d3-846c-7af40b51e8e6)
![image](https://github.com/user-attachments/assets/661acac6-2b01-4a5d-bbfc-97e0639b0f0b)

alinai commented:
@andyvanavery31 , please confirm, we want the text "Audio-Description:" or we want the icon for "Audio-Description:" in front of the link "Watch the audio-described version of this video."? 

Also, I could not see any comp with external link. Can you please if it looks good to you on ODE 871?

![image](https://github.com/user-attachments/assets/8295f62a-d821-4052-b265-b300d6f5a3fb)

![image](https://github.com/user-attachments/assets/1f691543-bdfa-47f3-8bbd-1932dace039c)

andyvanavery31 commented:
> @andyvanavery31 , please confirm, if we want the text "Audio-Description:" or we want the icon for "Audio-Description:" in front of the link "Watch the audio-described version of this video."?
> 
> Also, I could not see any comp with external link. Can you please confirm if it looks good to you on ODE 871?

@alinai The late change request from Alex is to use the icon followed by "Watch the audio-described version of this video." I updated the AC's to match this, and also added the Spanish version for testing too. Additionally, the external link icon appears due to legacy behavior and was not accounted for in the comp. This behavior is correct, and I added a note to confirm that in the ticket.

Cc: @dev-rana-publicis

dev-rana-publicis commented:
11/22 Push Fixes:
- Fixed test scenarios. We now have 4 test pages to test the audio-described link and other variations with the caption and audio-described link in English
  - Also added two Spanish test pages
- Spacing is fixed for when there is an audio-described link but no caption
- Changed audio-described link font-size to be 14px

andyvanavery31 commented:
- When there is no caption and only the AD link text, the total top spacing is 3.625px instead of 8px due to the -4.375px margin. Can this be removed?
- The tooltip text isn't appearing correctly

dev-rana-publicis commented:
11/25 Push Fixes:
- Fixed spacing on AD link when there is no caption
- Tooltip text is now showing the expected text

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_core/cgov_core.module:
```diff
@@ -700,7 +700,12 @@ function cgov_core_entity_build_defaults_alter(array &$build, EntityInterface $e
  */
 function cgov_core_field_widget_single_element_link_default_form_alter(&$element, FormStateInterface $form_state, $context) {
-  $element['uri']['#description'] = t('Enter the full URL starting with https://');
+  if ($element['#title'] == "Audio-Described Video Link") {
```
As discussed at standup, this needs to be moved to `cgov_video`, as this field does not exist without that. Also you could probably target the field name instead of the label? You should not have 20 different audio described fields to make this be generic. The current logic is because **ANY** external field, and you know we have a lot, removes things like autocomplete, and gets this overridden text.

Finally, order will matter. The code in the cgov_video _alter should fire after this. But just know that each hook is fired off one after the other based on the modules weight.

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_video/config/install/core.entity_view_display.media.cgov_video.video_display_medium_no_title.yml:
```diff
@@ -36,6 +38,18 @@ content:
     region: content
+  field_audio_described_link:
+    type: link
+    label: above
```
label should be hidden.

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_video/config/install/core.entity_view_display.media.cgov_video.video_display_large_title.yml:
```diff
@@ -36,6 +38,18 @@ content:
     region: content
+  field_audio_described_link:
+    type: link
+    label: above
```
hide

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_video/config/install/core.entity_view_display.media.cgov_video.video_display_large_no_title.yml:
```diff
@@ -36,6 +38,18 @@ content:
     region: content
+  field_audio_described_link:
+    type: link
+    label: above
```
hide

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_video/config/install/core.entity_view_display.media.cgov_video.video_display_medium_title.yml:
```diff
@@ -36,6 +38,18 @@ content:
     region: content
+  field_audio_described_link:
+    type: link
+    label: above
```
hide

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_video/config/install/core.entity_view_display.media.cgov_video.video_display_ncids_video.yml:
```diff
@@ -36,6 +38,18 @@ targetEntityType: media
 content:
+  field_audio_described_link:
+    type: link
+    label: above
```
hide

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_video/config/install/core.entity_view_display.media.cgov_video.video_display_small_no_title.yml:
```diff
@@ -36,6 +38,18 @@ content:
     region: content
+  field_audio_described_link:
+    type: link
+    label: above
```
hide

bryanpizzillo commented on docroot/profiles/custom/cgov_site/modules/custom/cgov_video/config/install/core.entity_view_display.media.cgov_video.video_display_small_title.yml:
```diff
@@ -36,6 +38,18 @@ content:
     region: content
+  field_audio_described_link:
+    type: link
+    label: above
```
hide

bryanpizzillo commented on docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/media/video/video_macros.twig:
```diff
@@ -34,6 +34,7 @@
           {{ content.field_caption }}
         </figcaption>
       {% endif %}
+      {{ content.field_audio_described_link }}
```
This is very wrong. `<figcaption>` needs to be the last thing in a figure.

bryanpizzillo replied:
(We went through this for images in videos in the 2 column...)

bryanpizzillo replied:
Well it is no longer in there, but yeah...I have new comments.

bryanpizzillo commented on docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/media/video/field--media--field-audio-described-link--cgov-video.html.twig:
```diff
@@ -0,0 +1,9 @@
+  {% set audio_described_text_1 = "Watch the"|t %}
+  {% set audio_described_text_2 = "audio-described version of this video"|t %}
+  <div class="cgdp-video__audio-described-link usa-prose">
+    <svg xmlns="http://www.w3.org/2000/svg" class="usa-icon" viewBox="0 0 24 24"><path d="M12.3 10.6H12v3.2h.3c.9 0 1.5-.8 1.5-1.6s-.7-1.6-1.5-1.6Zm5.1-4.7H6.6C4.6 5.9 3 7.5 3 9.5v5c0 2 1.6 3.6 3.6 3.6h10.8c2 0 3.6-1.6 3.6-3.6v-5c0-2-1.6-3.6-3.6-3.6Zm-7.2 9H8.8v-.7H6.7l-.5.7H4.3l4.2-5.8h1.7v5.8Zm2.5 0h-2.2V9.1h1.9c1.6 0 2.9 1.3 2.9 2.9s-1 2.9-2.6 2.9Zm3.2 0h-.5c.6-.7 1-1.7 1-2.7s-.5-2.3-1.3-3h.7c ...
```
This is supposed to be hidden from screen readers, right?

bryanpizzillo commented on docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/front-end/src/legacy/styles/_videos.scss:
```diff
@@ -122,3 +122,35 @@ span.video-title__heading {
+	// change the class used for the twig template based on if the page is display legacy vs.
+	// non-legacy content.
+	figcaption + .cgdp-video__audio-described-link {
+		padding-top: 0;
```
This will need to be fixed once the template changes. However, nice comment.

bryanpizzillo reviewed (changes requested):
A number of changes below

adrianacastaneda commented:
@dev-rana-publicis (Added this on Slack but adding to PR as well to help us track this issue better)

Had one small hiccup when testing. I tried to remove the title from the last video on [this page](https://ncigovcdode871.prod.acquia-sites.com/test/embedded-video-sizes-no-caption-ad) and the video disappeared from the editor. I can still see that it's there and if I double click on it and add a title back, the video preview returns but the video shouldn't disappear. Seems like a bug?

![Screenshot 2024-11-26 at 4 00 34 PM](https://github.com/user-attachments/assets/81873bc1-c71d-4671-826b-e77b30e8b820)

andyvanavery31 commented:
@dev-rana-publicis 
- Please update the AD link text to behave in an accessible way for screenreaders
- The test page for videos with [no captions nor AD links](https://ncigovcdode871.prod.acquia-sites.com/test/embedded-video-sizes-no-caption-ad) are missing the videos. They aren't in the body text of the CMS.
- While the padding looks technically correct when there is only the AD link, the text is slightly offset from where the padding is applied
![image](https://github.com/user-attachments/assets/f214f239-c366-4daa-9919-ef5155fce90e)

dev-rana-publicis commented:
12/9 Fixes:
- No caption and no audio-described link video now contains a valid, non-private youtube link
- Screenreader fixes:
  - Moved the audio-described link outside of the figure element
  - Added styling to apply the grey background to the audio-described link text and padding changes to apply the expected spacing as per the figma comps
  - Added an `aria-labelledby` on the figure's video, and an `id` on the figcaption for the screenreader to read the figcaption when active on the video
  - The figure element for the video will contain the class `.video__no-caption` if there is no caption present to add extra styling to achieve our desired spacing
- The figure element has 10px of marign-bottom, there is 17.5px of padding under the caption, there is 8px of padding ontop of the AD link text and our expected spacing between the caption and AD link is 16px.
  - 17.5px + 8px + X = 16px -> X = -9.5px
  - With the 10px margin-bottom we need to account for, I add a margin-top: -19.5px

KateMashkinaNIH commented:
Passed QA

bennettcc commented:
Received Fed approval on 12/17 from Anna/Alex

andyvanavery31 commented:
Videos without captions are failing accessibility because the caption is where the id for the video is set, and then the id is not initialized for videos without the caption.

bryanpizzillo commented on docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/media/video/video_macros.twig:
```diff
@@ -5,18 +5,21 @@
 
-{% macro video(name, content, showCaption = true, additionalClasses = "") %}
+{% macro video(name, content, media, showCaption = true, additionalClasses = "") %}
+  {% set videoID = 'media--' ~ media.id()|clean_unique_id %}
```
What does CTHP do where when it tries null.id()?

bryanpizzillo commented on docroot/profiles/custom/cgov_site/themes/custom/ncids_trans/templates/media/video/video_macros.twig:
```diff
@@ -28,12 +31,14 @@
         </figcaption>
       {% endif %}
     </figure>
+    {% if hasAudioDescribed %}
```
This should be in some sort of container... So if it must come after the figure, then you need to be elevating the class to have some sort of block element.

bryanpizzillo commented:
@andyvanavery31 I have concerns now that we are not making the legacy videos look like the new look/feel. This is partially making hacky css. Also all the video templates you broke out are getting touched in this anyway, so it is even harder to review this in light of "are these changes best for the future changes" aspect.

alinai commented:
@dev-rana-publicis ,  video on the page level is not showing the AD link. Please see the note below from @andyvanavery31:

![image](https://github.com/user-attachments/assets/820d5e32-c23c-4675-9440-ff8333b4c0f3)

KateMashkinaNIH commented:
looks good
